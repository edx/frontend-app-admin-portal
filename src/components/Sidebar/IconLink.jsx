import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import classNames from 'classnames';
import { Icon } from '@edx/paragon';

const IconLink = props => (
  <NavLink
    className="nav-link text-left rounded-0"
    to={props.to}
  >
    <Icon
      className={classNames([
        'fa',
        props.iconClassName,
        {
          'mr-2': props.isExpanded,
        },
      ])}
      screenReaderText={!props.isExpanded ? props.title : undefined}
    />
    {props.isExpanded && props.title}
  </NavLink>
);

IconLink.defaultProps = {
  iconClassName: null,
  isExpanded: false,
};

IconLink.propTypes = {
  title: PropTypes.string.isRequired,
  to: PropTypes.string.isRequired,
  iconClassName: PropTypes.string,
  isExpanded: PropTypes.bool,
};

export default IconLink;
