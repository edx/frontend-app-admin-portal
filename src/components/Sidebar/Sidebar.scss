#sidebar {
    width: 250px;
    padding-top: 50px;
    position: relative;
    float: left;
    top: 0;
    left: 0;
    height: 100vh;
    overflow: hidden;
    z-index: 9999;
    background: #fff;
    transition: all 0.3s;

    overflow-y: scroll;

    -moz-box-shadow:    0 .125rem .25rem rgba(0,0,0,.075);
    -webkit-box-shadow: 0 .125rem .25rem rgba(0,0,0,.075);
    box-shadow:         0 .125rem .25rem rgba(0,0,0,.075);

    .btn-toggle {
        position: fixed;
        left: 10px;
        top: 5px;
        background: transparent;
    }

    .sidebar-items {
        height: 100%;
        overflow: auto;
    }

    &.collapsed {
        margin-left: -250px;
    }
}

.sidebar-collapsed {
    position: fixed;
    z-index: 999;
}
