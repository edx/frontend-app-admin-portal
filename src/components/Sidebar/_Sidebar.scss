.sidebar {
  background: $gray-100;
  position: absolute;
  top: 0;
  left: 0;
  z-index: $zindex-fixed;

  &.expanded {
    width: auto;

    &.has-shadow {
      box-shadow: 6px 0 10px -6px rgba(0, 0, 0, 0.25);
    }
  }

  .sidebar-content {
    .nav {
      position: sticky;
      top: 0;

      .nav-link {
        &.active {
          background: $primary-200;

          &:hover,
          &:focus {
            background: $primary-300;
          }
        }

        &:hover,
        &:focus {
          background: $gray-200;
        }
      }
    }
  }
}
