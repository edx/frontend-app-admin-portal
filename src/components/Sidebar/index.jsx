import React from 'react';
import PropTypes from 'prop-types';
import SidebarToggle from '../SidebarToggle';
import SidebarItem from '../SidebarItem';

import './Sidebar.scss';

function Sidebar(props) {
  const {
    sidebarEnabled, sidebarExpanded, sidebarData, toggleSidebar, itemsToShow,
  } = props;

  if (!sidebarEnabled) {
    return null;
  }

  return sidebarExpanded ? (
    <div id="sidebar">
      <SidebarToggle expanded={sidebarExpanded} toggleSidebar={toggleSidebar} />
      <div className="sidebar-items">
        {
          sidebarData.map(item => (
            <SidebarItem itemData={item} key={item.category} itemsToShow={itemsToShow} />
          ))
        }
      </div>
    </div>
  ) :
    (
      <div id="sidebar" className="collapsed">
        <SidebarToggle expanded={sidebarExpanded} toggleSidebar={toggleSidebar} />
      </div>
    );
}

Sidebar.defaultProps = {
  sidebarExpanded: false,
  sidebarEnabled: false,
  sidebarData: [],
  itemsToShow: 5,
  toggleSidebar: () => {},
};

Sidebar.propTypes = {
  sidebarExpanded: PropTypes.bool,
  sidebarEnabled: PropTypes.bool,
  sidebarData: PropTypes.array, // eslint-disable-line react/forbid-prop-types
  itemsToShow: PropTypes.number,
  toggleSidebar: PropTypes.func,
};

export default Sidebar;
