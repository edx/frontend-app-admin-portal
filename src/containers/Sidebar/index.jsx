import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import Sidebar from '../../components/Sidebar';

import {
  expandSidebar,
  collapseSidebar,
} from '../../data/actions/sidebar';

const mapStateToProps = state => ({
  isExpanded: state.sidebar.isExpanded,
  isExpandedByToggle: state.sidebar.isExpandedByToggle,
});

const mapDispatchToProps = dispatch => ({
  expandSidebar: () => dispatch(expandSidebar()),
  collapseSidebar: (usingToggle = false) => dispatch(collapseSidebar(usingToggle)),
});

export default withRouter(connect(
  mapStateToProps,
  mapDispatchToProps,
  null,
  { withRef: true },
)(Sidebar));
