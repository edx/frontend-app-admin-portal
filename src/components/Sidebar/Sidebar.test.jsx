import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Sidebar from './index';

const SIDEBAR_DATA = [
  {
    category: 'Catalogs',
    showAllButton: true,
    action: {
      type: 'BTN_LINK',
      title: 'Create New Catalog',
      href: 'https://reactjs.org/',
    },
    items: [
      {
        type: 'TXT_LINK',
        title: 'JS Catalog',
        href: 'https://js.org/',
      }, {
        type: 'TXT_LINK',
        title: 'PY Catalog',
        href: 'https://www.python.org/',
      },
    ],
  }, {
    category: 'Management',
    showAllButton: false,
    items: [
      {
        type: 'TXT_LINK',
        title: 'Peoeple & Groups',
        href: 'https://edx.org/',
      },
    ],
  },
];

describe('<Sidebar />', () => {
  let wrapper;

  it('will not render if disabled', () => {
    const tree = renderer
      .create((
        <Sidebar
          sidebarExpanded
          sidebarEnabled={false}
          sidebarData={SIDEBAR_DATA}
          toggleSidebar={() => {}}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when expanded', () => {
    const tree = renderer
      .create((
        <Sidebar
          sidebarExpanded
          sidebarEnabled
          sidebarData={SIDEBAR_DATA}
          toggleSidebar={() => {}}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when collapsed', () => {
    const tree = renderer
      .create((
        <Sidebar
          sidebarExpanded={false}
          sidebarEnabled
          sidebarData={SIDEBAR_DATA}
          toggleSidebar={() => {}}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has `show all` link working correctly', () => {
    wrapper = mount((
      <Sidebar
        sidebarExpanded
        sidebarEnabled
        sidebarData={SIDEBAR_DATA}
        itemsToShow={1}
        toggleSidebar={() => {}}
      />
    ));

    // verify initial data
    expect(wrapper.find('.sidebar-item-wrapper.catalogs').find('.sidebar-item-header button').text())
      .toEqual('show all');
    expect(wrapper.find('.sidebar-item-wrapper.catalogs').find('li.sidebar-item-child').length).toEqual(2);
    expect(wrapper.find('.sidebar-item-wrapper.management').find('li.sidebar-item-child').length).toEqual(1);

    // verify things after click on `show all` button
    wrapper.find('.sidebar-item-wrapper.catalogs').find('.sidebar-item-header button').simulate('click');
    expect(wrapper.find('.sidebar-item-wrapper.catalogs').find('li.sidebar-item-child').length).toEqual(3);

    // click on `show less` and verify that things reset to initial data
    expect(wrapper.find('.sidebar-item-wrapper.catalogs').find('.sidebar-item-header button').text())
      .toEqual('show less');
    wrapper.find('.sidebar-item-wrapper.catalogs').find('.sidebar-item-header button').simulate('click');
    expect(wrapper.find('.sidebar-item-wrapper.management').find('li.sidebar-item-child').length).toEqual(1);
  });
});
