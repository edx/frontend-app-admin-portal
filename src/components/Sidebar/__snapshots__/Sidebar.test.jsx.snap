// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> renders correctly when collapsed 1`] = `
<div
  className="collapsed"
  id="sidebar"
>
  <button
    className="btn btn-toggle"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span>
      <span>
        <span
          aria-hidden={true}
          className="fa fa-bars"
          id="Icon1"
        />
      </span>
    </span>
  </button>
</div>
`;

exports[`<Sidebar /> renders correctly when expanded 1`] = `
<div
  id="sidebar"
>
  <button
    className="btn btn-toggle"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span>
      <span>
        <span
          aria-hidden={true}
          className="fa fa-times"
          id="Icon1"
        />
      </span>
    </span>
  </button>
  <div
    className="sidebar-items"
  >
    <section
      className="sidebar-item-wrapper catalogs"
    >
      <div
        className="sidebar-item-header"
      >
        <h5
          id="sidebar-item-catalogs"
        >
          Catalogs
        </h5>
        <button
          className="btn btn-link"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          show all
        </button>
      </div>
      <nav
        aria-labelledby="sidebar-item-catalogs"
        className="sidebar-item-childs"
      >
        <ul>
          <li
            className="sidebar-item-child"
          >
            <a
              href="https://js.org/"
            >
              JS Catalog
            </a>
          </li>
          <li
            className="sidebar-item-child"
          >
            <a
              href="https://www.python.org/"
            >
              PY Catalog
            </a>
          </li>
          <li
            className="sidebar-item-child btn-sidebar"
          >
            <a
              href="https://reactjs.org/"
              onClick={[Function]}
              target="_self"
            >
              <span>
                <span>
                  <span
                    aria-hidden={true}
                    className="content-icon"
                    id="Icon1"
                  />
                </span>
                Create New Catalog
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </section>
    <section
      className="sidebar-item-wrapper management"
    >
      <div
        className="sidebar-item-header"
      >
        <h5
          id="sidebar-item-management"
        >
          Management
        </h5>
      </div>
      <nav
        aria-labelledby="sidebar-item-management"
        className="sidebar-item-childs"
      >
        <ul>
          <li
            className="sidebar-item-child"
          >
            <a
              href="https://edx.org/"
            >
              Peoeple & Groups
            </a>
          </li>
        </ul>
      </nav>
    </section>
  </div>
</div>
`;

exports[`<Sidebar /> will not render if disabled 1`] = `null`;
